analyzer/protocol/openvpn/openvpn.spicy

```diff
@@ -0,0 +1,181 @@
+	};
+
+	switch ( self.message_type.opcode ) {
+		Opcode::P_CONTROL_HARD_RESET_CLIENT_V1 -> control_hard_reset_client_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_CONTROL_HARD_RESET_SERVER_V1 -> control_hard_reset_server_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_CONTROL_SOFT_RESET_V1 -> control_soft_reset_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_CONTROL_V1 -> control_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_ACK_V1 -> ack_v1: AckMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_DATA_V1 -> data_v1: DataMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_CONTROL_HARD_RESET_SERVER_V2 -> control_hard_reset_server_v2: ControlMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_CONTROL_HARD_RESET_CLIENT_V2 -> control_hard_reset_client_v2: ControlMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_DATA_V2 -> data_v2: DataMessage(self.message_type.opcode, self.message_type.key_id, True, share);
+	};
+};
+
+public type OpenVPNRecordsHMAC = unit {
+	%context=Share;
+	records : OpenVPNRecordHMAC(self.context())[];
+};
+
+public type OpenVPNRecordHMAC = unit(inout share: Share) {
+	message_type: bitfield(8) {
+		opcode: 3..7 &convert=Opcode($$);
+		key_id: 0..2;
+	};
+
+	switch ( self.message_type.opcode ) {
+		Opcode::P_CONTROL_HARD_RESET_CLIENT_V1 -> control_hard_reset_client_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, True, share);
+		Opcode::P_CONTROL_HARD_RESET_SERVER_V1 -> control_hard_reset_server_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, True, share);
+		Opcode::P_CONTROL_SOFT_RESET_V1 -> control_soft_reset_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, True, share);
+		Opcode::P_CONTROL_V1 -> control_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, True, share);
+		Opcode::P_ACK_V1 -> ack_v1: AckMessage(self.message_type.opcode, self.message_type.key_id, True, share);
+		Opcode::P_DATA_V1 -> data_v1: DataMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_CONTROL_HARD_RESET_SERVER_V2 -> control_hard_reset_server_v2: ControlMessage(self.message_type.opcode, self.message_type.key_id, True, share);
+		Opcode::P_CONTROL_HARD_RESET_CLIENT_V2 -> control_hard_reset_client_v2: ControlMessage(self.message_type.opcode, self.message_type.key_id, True, share);
+		Opcode::P_DATA_V2 -> data_v2: DataMessage(self.message_type.opcode, self.message_type.key_id, True, share);
+	};
+};
+
+public type OpenVPNRecordsTCP = unit {
+	%context=Share;
+	records : OpenVPNRecordTCP(self.context())[];
+};
+
+public type OpenVPNRecordTCP = unit(inout share: Share) {
```

Comment:
To avoid repeating the switch statement here, would it work to do this:

```
public type OpenVPNRecordTCP = unit(inout share: Share) {
    packet_length : uint16;
    record: OpenVPNRecord &size=self.packet_length - 1;
};
```
?

Reply:
I think it will.  I'll have to test.

Reply:
This is the only suggestion you made I can't get to work right.  I added that change but the tests fail and I can't figure out why.

Reply:
I left a TODO: in the code for this.

Reply:
I took a quick look at this and am also not sure why that isn't working. I'll merge as is for  now and will create a ticket to look into it.